packages/react-sdk/src/components/BackgroundFilters/BackgroundFilters.tsx

```diff
@@ -197,166 +197,124 @@ export const BackgroundFiltersProvider = (
-  // it should be called with the filtered MediaStream as an argument.
-  const signalFilterReadyRef =
-    useRef<(value: MediaStream | PromiseLike<MediaStream>) => void>();
-
-  const [mediaStream, setMediaStream] = useState<MediaStream>();
-  const unregister = useRef<Promise<void>>();
+  const { children, start } = useRenderer(props.tfLite);
+  const { backgroundFilter } = useBackgroundFilters();
+
   useEffect(() => {
-    if (!call || !backgroundFilter) return;
-    const cleanup = () => {
-      signalFilterReadyRef.current = undefined;
-      setMediaStream(undefined);
-    };
-    const register = (unregister.current || Promise.resolve()).then(() =>
-      call.camera.registerFilter(async (ms) => {
-        return new Promise<MediaStream>((resolve) => {
-          setMediaStream(ms);
-          signalFilterReadyRef.current = resolve;
-        });
-      }, cleanup),
-    );
+    if (!call || !backgroundFilter) {
+      return;
+    }
 
+    const { unregister } = call.camera.registerFilter((ms) => start(ms));
     return () => {
-      unregister.current = register
-        .then((unregisterFilter) => unregisterFilter())
-        .catch((err) => console.error('Failed to unregister filter', err));
+      unregister();
     };
-  }, [backgroundFilter, call]);
+  }, [backgroundFilter, call, start]);
 
-  const [isPlaying, setIsPlaying] = useState(false);
-  useEffect(() => {
-    if (!mediaStream || !videoRef) return;
-    const handleOnPlay = () => {
-      const [track] = mediaStream.getVideoTracks();
-      if (!track) return;
-      const { width: w = 0, height: h = 0 } = track.getSettings();
-      setWidth(w);
-      setHeight(h);
-      setIsPlaying(true);
-    };
-    videoRef.addEventListener('play', handleOnPlay);
-    videoRef.srcObject = mediaStream;
-    videoRef.play().catch((err) => {
-      console.error('Failed to play video', err);
-    });
-    return () => {
-      videoRef.removeEventListener('play', handleOnPlay);
-      videoRef.srcObject = null;
-      setIsPlaying(false);
-    };
-  }, [mediaStream, videoRef]);
+  return children;
+};
 
-  useEffect(() => {
-    const resolveFilter = signalFilterReadyRef.current;
-    if (!canvasRef || !resolveFilter) return;
+const useRenderer = (tfLite: TFLite) => {
+  const { backgroundFilter, backgroundBlurLevel, backgroundImage } =
+    useBackgroundFilters();
+  const videoRef = useRef<HTMLVideoElement>(null);
+  const canvasRef = useRef<HTMLCanvasElement>(null);
+  const bgImageRef = useRef<HTMLImageElement>(null);
+  const [videoSize, setVideoSize] = useState<{ width: number; height: number }>(
+    {
+      width: 1920,
+      height: 1080,
+    },
+  );
 
-    const filter = canvasRef.captureStream();
-    resolveFilter(filter);
-    return () => {
-      disposeOfMediaStream(filter);
-    };
-  }, [canvasRef]);
+  const start = useCallback(
+    (ms: MediaStream) => {
+      return new Promise<[MediaStream, () => void]>((resolve, reject) => {
```

Comment:
That's the whole point of this PR: to rewrite background filter pipeline as a single linear function, instead of an effects-based logic.